### PR TITLE
If condition variables support

### LINKED ISSUE #344: False error in conditional block
Error message `Specify == or !=` is shown for valid condition expression if is not an equality check.

```
var cond = true
if cond 

endif
```

### PR DESCRIPTION
**Pull request by PolinaShneider**
Closes #344 

Added support for `if` conditions declared as variables

image

### CODE REVIEW

**Comment by VeselovAlex** [comment]
What about conditions like `someList.length > 10` or `someThing.values('field').includes('sth')` ?

File: server/src/validator.ts [hunk 1]
```diff
@@ -684,6 +684,46 @@ export class Validator {
...
+            name = varRegex.exec(line) ? varRegex.exec(line)[1] : null;
+        }
+
+        // if `var condition = value` not found, it must be inline syntax
+        if (!name && !correctInlineIf) {
+            this.result.push(createDiagnostic(this.foundKeyword.range, `Specify == or !=`));
+        } else if (name !== nameInCondition) {
+            // if `var condition = value` was found but variable name doesn't match
+            this.result.push(createDiagnostic(this.foundKeyword.range,
+                `var '${nameInCondition}' used in 'if' condition not found`));
```
Comment (annstriganova) [inline server/src/validator.ts]: Move to `messageUtil`.

File: server/src/validator.ts [hunk 2]
```diff
@@ -684,6 +684,46 @@ export class Validator {
...
+
+        // search for `var condition = value`
+        while (!name && line !== null) {
+            line = this.config.getLine(--num);
+            name = varRegex.exec(line) ? varRegex.exec(line)[1] : null;
+        }
+
+        // if `var condition = value` not found, it must be inline syntax
+        if (!name && !correctInlineIf) {
+            this.result.push(createDiagnostic(this.foundKeyword.range, `Specify == or !=`));
```
Comment (annstriganova) [inline server/src/validator.ts]: Move to `messageUtil`.

File: server/src/validator.ts [hunk 3]
```diff
@@ -684,6 +684,46 @@ export class Validator {
         return setting;
     }
 
+    /**
+     * Checks `if` condition syntax
+     */
+    private handleIf(): void {
```
Comment (annstriganova) [inline server/src/validator.ts]: Can it be moved to `KeywordHandler`?

File: server/src/keywordHandler.ts [hunk 4]
```diff
@@ -42,6 +42,25 @@ export class KeywordHandler {
...
+
+        try {
+            Function(`return ${ifCondition}`);
+        } catch (err) {
+            const range = ifCondition ? createRange(
+                line.indexOf(ifCondition),
+                ifCondition.length,
+                foundKeyword.range.start.line
+            ) : foundKeyword.range;
+            this.diagnostics.push(createDiagnostic(range, "Wrong if condition"));
```
Comment (VeselovAlex) [inline server/src/keywordHandler.ts]: Provide information about what is actually wrong, if possible

File: server/src/keywordHandler.ts [hunk 5]
```diff
@@ -42,6 +42,25 @@ export class KeywordHandler {
         }
     }
 
+    /**
+     * Checks `if` condition syntax
+     */
+    public handleIf(line: string, foundKeyword: TextRange): void {
+        const ifConditionRegex: RegExp = /^[\s].*if\s*(.*)/;
+        const ifCondition = ifConditionRegex.exec(line) ? ifConditionRegex.exec(line)[1] : null;
```
Comment (VeselovAlex) [inline server/src/keywordHandler.ts]: Should we just return if it is not if expression?

Comment (VeselovAlex) [reply server/src/keywordHandler.ts]: Aren't it this condition checked before `handleIf` call?

**Comment by annstriganova** [comment]
Expected error in this case, but there is no one:
```elm
if
endif
```

**Comment by PolinaShneider** [comment]
@annstriganova it is interpreted as `""` which is equivalent to false. I will add a check for this case

File: server/src/keywordHandler.ts [hunk 6]
```diff
@@ -39,6 +39,13 @@ export class KeywordHandler {
         const ifConditionRegex: RegExp = /^[\s].*if\s*(.*)/;
         const ifCondition = ifConditionRegex.exec(line)[1];
 
+        if (ifCondition.trim() === "") {
+            this.diagnostics.push(
+                createDiagnostic(foundKeyword.range, "If condition mustn't be empty")
```
Comment (annstriganova) [inline server/src/keywordHandler.ts]: `mustn't` -> `can not`

**Pull request merged**